File: src/xpressai/core/activity.py
"""Activity logging for XpressAI.

Tracks events across the system for observability and debugging.
"""

from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Any
import json
import logging

from xpressai.memory.database import Database

logger = logging.getLogger(__name__)


class EventType(str, Enum):
    """Types of activity events."""

    # Task events
    TASK_CREATED = "task.created"
    TASK_STARTED = "task.started"
    TASK_COMPLETED = "task.completed"
    TASK_FAILED = "task.failed"
    TASK_WAITING = "task.waiting"

    # Agent events
    AGENT_STARTED = "agent.started"
    AGENT_STOPPED = "agent.stopped"
    AGENT_ERROR = "agent.error"

    # Tool events
    TOOL_CALLED = "tool.called"
    TOOL_COMPLETED = "tool.completed"
    TOOL_FAILED = "tool.failed"

    # System events
    SYSTEM_STARTUP = "system.startup"
    SYSTEM_SHUTDOWN = "system.shutdown"

    # User events
    USER_MESSAGE = "user.message"


@dataclass
class ActivityEvent:
    """An activity event.

    Attributes:
        id: Event ID
        timestamp: When the event occurred
        event_type: Type of event
        agent_id: Related agent (if any)
        data: Additional event data
        session_id: Related session (if any)
    """

    id: int
    timestamp: datetime
    event_type: EventType
    agent_id: str | None
    data: dict[str, Any]
    session_id: str | None


class ActivityManager:
    """Manages activity logging and querying.

    Provides methods to log events and retrieve activity history.
    """

    def __init__(self, db: Database):
        """Initialize activity manager.

        Args:
            db: Database instance
        """
        self.db = db

    async def log(
        self,
        event_type: EventType,
        agent_id: str | None = None,
        data: dict[str, Any] | None = None,
        session_id: str | None = None,
    ) -> ActivityEvent:
        """Log an activity event.

        Args:
            event_type: Type of event
            agent_id: Related agent
            data: Additional event data
            session_id: Related session

        Returns:
            Created event
        """
        with self.db.connect() as conn:
            cursor = conn.execute(
                """
                INSERT INTO activity_logs (agent_id, event_type, event_data, session_id)
                VALUES (?, ?, ?, ?)
            """,
                (
                    agent_id,
                    event_type.value,
                    json.dumps(data) if data else None,
                    session_id,
                ),
            )
            event_id = cursor.lastrowid

            row = conn.execute(
                "SELECT * FROM activity_logs WHERE id = ?", (event_id,)
            ).fetchone()

            event = self._row_to_event(row)
            logger.debug(f"Logged activity: {event_type.value} - {data}")
            return event

    async def get_recent(self, limit: int = 50) -> list[ActivityEvent]:
        """Get recent activity events.

        Args:
            limit: Maximum number of events to return

        Returns:
            List of events, most recent first
        """
        with self.db.connect() as conn:
            rows = conn.execute(
                """
                SELECT * FROM activity_logs
                ORDER BY timestamp DESC
                LIMIT ?
            """,
                (limit,),
            ).fetchall()

            return [self._row_to_event(row) for row in rows]

    async def get_by_agent(
        self, agent_id: str, limit: int = 50
    ) -> list[ActivityEvent]:
        """Get activity events for a specific agent.

        Args:
            agent_id: Agent ID to filter by
            limit: Maximum number of events

        Returns:
            List of events for the agent
        """
        with self.db.connect() as conn:
            rows = conn.execute(
                """
                SELECT * FROM activity_logs
                WHERE agent_id = ?
                ORDER BY timestamp DESC
                LIMIT ?
            """,
                (agent_id, limit),
            ).fetchall()

            return [self._row_to_event(row) for row in rows]

    async def get_by_type(
        self, event_type: EventType, limit: int = 50
    ) -> list[ActivityEvent]:
        """Get activity events of a specific type.

        Args:
            event_type: Event type to filter by
            limit: Maximum number of events

        Returns:
            List of events of the given type
        """
        with self.db.connect() as conn:
            rows = conn.execute(
                """
                SELECT * FROM activity_logs
                WHERE event_type = ?
                ORDER BY timestamp DESC
                LIMIT ?
            """,
                (event_type.value, limit),
            ).fetchall()

            return [self._row_to_event(row) for row in rows]

    async def clear_old(self, days: int = 30) -> int:
        """Clear activity logs older than specified days.

        Args:
            days: Number of days to keep

        Returns:
            Number of events deleted
        """
        with self.db.connect() as conn:
            cursor = conn.execute(
                """
                DELETE FROM activity_logs
                WHERE timestamp < datetime('now', ?)
            """,
                (f"-{days} days",),
            )
            count = cursor.rowcount
            logger.info(f"Cleared {count} old activity logs")
            return count

    def _row_to_event(self, row) -> ActivityEvent:
        """Convert a database row to ActivityEvent."""
        data = {}
        if row["event_data"]:
            try:
                data = json.loads(row["event_data"])
            except json.JSONDecodeError:
                pass

        return ActivityEvent(
            id=row["id"],
            timestamp=datetime.fromisoformat(row["timestamp"])
            if row["timestamp"]
            else datetime.now(),
            event_type=EventType(row["event_type"]),
            agent_id=row["agent_id"],
            data=data,
            session_id=row["session_id"],
        )

File: crates/xpressclaw-core/src/activity.rs
use std::sync::Arc;

use serde::{Deserialize, Serialize};
use tracing::debug;

use crate::db::Database;
use crate::error::{Error, Result};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActivityEvent {
    pub id: i64,
    pub timestamp: String,
    pub event_type: String,
    pub agent_id: Option<String>,
    /// Serialized as `event_data` to match DB column and frontend API contract.
    #[serde(rename = "event_data")]
    pub data: Option<serde_json::Value>,
    pub session_id: Option<String>,
}

/// Activity logger for observability.
pub struct ActivityManager {
    db: Arc<Database>,
}

impl ActivityManager {
    pub fn new(db: Arc<Database>) -> Self {
        Self { db }
    }

    pub fn log(
        &self,
        event_type: &str,
        agent_id: Option<&str>,
        data: Option<&serde_json::Value>,
        session_id: Option<&str>,
    ) -> Result<ActivityEvent> {
        let data_json = data.map(|d| d.to_string());
        let conn = self.db.conn();

        conn.execute(
            "INSERT INTO activity_logs (agent_id, event_type, event_data, session_id) VALUES (?1, ?2, ?3, ?4)",
            rusqlite::params![agent_id, event_type, data_json, session_id],
        )?;

        let id = conn.last_insert_rowid();
        let mut stmt = conn.prepare("SELECT * FROM activity_logs WHERE id = ?1")?;
        let event = stmt
            .query_row([id], |row| Ok(row_to_event(row)))
            .map_err(|e| Error::Database(e.to_string()))??;

        debug!(event_type, ?agent_id, "logged activity");
        Ok(event)
    }

    pub fn get_recent(&self, limit: i64) -> Result<Vec<ActivityEvent>> {
        let conn = self.db.conn();
        let mut stmt =
            conn.prepare("SELECT * FROM activity_logs ORDER BY timestamp DESC LIMIT ?1")?;

        let events = stmt
            .query_map([limit], |row| Ok(row_to_event(row)))
            .map_err(|e| Error::Database(e.to_string()))?
            .filter_map(|r| r.ok())
            .filter_map(|r| r.ok())
            .collect();

        Ok(events)
    }

    pub fn get_by_agent(&self, agent_id: &str, limit: i64) -> Result<Vec<ActivityEvent>> {
        let conn = self.db.conn();
        let mut stmt = conn.prepare(
            "SELECT * FROM activity_logs WHERE agent_id = ?1 ORDER BY timestamp DESC LIMIT ?2",
        )?;

        let events = stmt
            .query_map(rusqlite::params![agent_id, limit], |row| {
                Ok(row_to_event(row))
            })
            .map_err(|e| Error::Database(e.to_string()))?
            .filter_map(|r| r.ok())
            .filter_map(|r| r.ok())
            .collect();

        Ok(events)
    }
}

fn row_to_event(row: &rusqlite::Row) -> Result<ActivityEvent> {
    let data_str: Option<String> = row.get("event_data")?;
    let data = data_str
        .as_deref()
        .and_then(|s| serde_json::from_str(s).ok());

    Ok(ActivityEvent {
        id: row.get("id")?,
        timestamp: row.get("timestamp")?,
        event_type: row.get("event_type")?,
        agent_id: row.get("agent_id")?,
        data,
        session_id: row.get("session_id")?,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_log_and_retrieve() {
        let db = Arc::new(Database::open_memory().unwrap());
        let mgr = ActivityManager::new(db);

        let data = serde_json::json!({"task_id": "t1", "title": "Test"});
        let event = mgr
            .log("task.created", Some("atlas"), Some(&data), None)
            .unwrap();

        assert_eq!(event.event_type, "task.created");
        assert_eq!(event.agent_id.as_deref(), Some("atlas"));

        let recent = mgr.get_recent(10).unwrap();
        assert_eq!(recent.len(), 1);

        let by_agent = mgr.get_by_agent("atlas", 10).unwrap();
        assert_eq!(by_agent.len(), 1);
    }
}
